Replace message parsing with throwing more specific exceptions

vyasr opened the pull request:
## Description
<!-- Provide a standalone description of changes in this PR. -->
<!-- Reference any issues closed by this PR with "closes #1234". -->
<!-- Note: The pull request title will be included in the CHANGELOG. -->
This PR identifies places where cuDF Python is parsing exception messages from libcudf in order to throw an appropriate Python exception and modifies the associated libcudf error macros (`CUDF_EXPECTS` or `CUDF_FAIL`) to throw a more descriptive exception. In order to fully support this behavior with arbitrary libcudf exceptions, this PR also introduces a custom Cython exception handler that can be extended to catch and handle any new type of exception thrown by libcudf. To cases where issues with the dtype of an argument is treated as a TypeError rather than a ValueError in pandas, a new exception type `dtype_error : public logic_error` is defined and mapped to a TypeError in Python. This PR does not aim to exhaustively improve the choice of thrown exceptions throughout libcudf, only in places that are immediately relevant to removing message parsing in cudf Python.

Resolves #10632 

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

vyasr commented:
The longer term plan here is to update all of libcudf's exception handling to throw something more useful than a `cudf::logic_error`. I'd like to use this PR as a test case for that to sort through some questions. There are a few issues that need to be discussed to move this work forward:
- @jlowe @revans2 I'm not sure how changes like this impact Java. My understanding is that all exceptions get mapped to [CudfException](https://github.com/rapidsai/cudf/blob/branch-23.02/java/src/main/java/ai/rapids/cudf/CudfException.java) or [CudaException](https://github.com/rapidsai/cudf/blob/branch-23.02/java/src/main/java/ai/rapids/cudf/CudaException.java) or so according to the logic defined in [jni_utils.hpp](https://github.com/rapidsai/cudf/blob/branch-23.02/java/src/main/native/include/jni_utils.hpp). Based on what I see there, any exception that is not one of the "known types" in `CATCH_STD` (basically an rmm OOM or a CUDA error) gets mapped to a `CudfException`. If that is true, then AFAICT changing the exception types in C++ would have no effect at all on the JNI in its current state because even any `std::*` exception would just get mapped to a `CudfException`. Is that correct? 
- @jlowe @revans2 Aside from making sure that this change is non-breaking for the JNI, are there ways in which this change to libcudf could be more useful in the Java plugin?
- @shwina @jrhemstad currently this PR is incomplete and does not remove all cases where cuDF Python is parsing the exception message. Many of the remaining examples are in [lists.py](https://github.com/rapidsai/cudf/blob/branch-23.02/python/cudf/cudf/core/column/lists.py). The problem is that there are cases where the [Cython/C++ exception mapping](http://docs.cython.org/en/latest/src/userguide/wrapping_CPlusPlus.html#exceptions) is not just incomplete, but rather disagrees with what pandas (and therefore cudf) wants to throw. The most common example is that there are places where pandas throws a TypeError based on the dtype of an input. [Here is a case where libcudf is checking the input parameters for some equality condition](https://github.com/rapidsai/cudf/blob/branch-23.02/cpp/src/lists/contains.cu#L312). I would expect this to throw a `std::invalid_argument`. However, in Python a `std::invalid_argument` is translated to a ValueError, whereas [what cudf actually wants to throw here is a TypeError](https://github.com/rapidsai/cudf/blob/branch-23.02/python/cudf/cudf/core/column/lists.py#L408). I am not sure what the best way to handle this kind of issue is. If there is only one way that a particular exception can be thrown by a libcudf API, then it's not a problem; we can just catch a ValueError and reraise it as a TypeError. The much more troublesome case is when there are multiple parameters, each of which could be invalid, so a `std::invalid_exception` from libcudf could be coming from any of them. We could try defining additional exception types in libcudf that are more specific in order to differentiate these cases, but I suspect that some of these case are not fundamentally different, they're just viewed as different by the pandas API. As a result, we would be leaking pandas details into the types defined in libcudf.

revans2 commented:
For the Java side of things I see the general pattern to be that each time you add in a new exception type in C++, it would be good to add a corresponding subclass of CudfException, and set up a mapping between the two in CATCH_STD.  This would maintain backwards compatibility with all existing code, but allow us to add in much more specific exception types.

If you want to you can also keep the code as is for now and we can handle adding in the mappings we want to add in. The important part for us is to be able to distinguish between different types of errors so we can start to understand if there are better ways to recover from the error besides just crashing and letting Spark retry it.

shwina commented:
> As a result, we would be leaking pandas details into the types defined in libcudf. Any thoughts on the best way to handle this?

It sounds like inevitably, there's going to be cases where libcudf will throw the wrong kind of error. I'd say that it's a success if we're able to get rid of exception message parsing for the most part, while still having a few stragglers.

shwina commented:
Retargeted to 23.04

vyasr commented:
@shwina it turns out I was wrong. At least based on the current usage, there is a fairly specific class of cases where I was expecting to throw a `std::invalid_argument` but pandas wants to see a `TypeError`: cases where the validity issue is specifically due to the data type of the input. To account for this case, I've added a new `dtype_error` in libcudf that captures this more precisely, and I introduced a custom exception handler in our Cython layer to handle it.

vyasr commented:
@revans2 I introduced one new exception type in this PR and I added a suitable additional mapping for it in `CATCH_STD`. Let me know if there is anything additional that you would like to see.

vyasr commented:
For Python reviewers: once everyone has approved, I plan to run a batch sed update to change `except +` to `except +cudf_exception_handler` throughout our Cython code, and I will update our [developer documentation for exception handling](https://github.com/rapidsai/cudf/blob/branch-23.04/docs/cudf/source/developer_guide/contributing_guide.md#handling-libcudf-exceptions) accordingly.

shwina commented on python/cudf/cudf/_lib/cpp/exception_handler.hpp:
```diff
@@ -0,0 +1,80 @@
+/*
```
The `cudf/_lib/cpp` directory contains the Cython definitions ("declarations") for `libcudf` types and functions.

Perhaps we can place this (and the corresponding `.pxd` file) in `cudf/_lib/exception_handler.hpp` (and `cudf/_lib/exception_handler.pxd`) instead?

Alternately, I'd also be fine with just inlining this bit of C++ in `cudf/_lib_exception_handler.pxd`

vyasr replied:
Yeah, I wasn't sure what the place was for this given the C++. Personally I prefer avoiding inline C++ in pxd files, but I can go that route if you like it. IMO it will be familiar to an even smaller subset of devs, and no editors are capable of syntactically or semantically parsing it as anything other than a comment. Moving it to `_lib` is also fine with me; it seemed somewhat appropriate to put in `cpp` since `exception_handler.pxd` is now also just a file of Cython "declarations" rather than new definitions, but on further consideration I think `_lib` is more appropriate too.

shwina replied:
OK let's move them both to `_lib` (and keep them separate)

vyasr replied:
I've made this change, but one thing to note is that it does make for some slightly awkward circularity. We are now importing the handler from `_lib` into all the pxd files in `_lib/cpp` (because the latter is where we actually expose C++ APIs that need to be decorated with +cudf_exception_handler). Do you still like that outcome?

shwina replied:
I think that's fine

ttnghia commented on cpp/include/cudf/lists/combine.hpp:
```diff
@@ -80,8 +80,8 @@ std::unique_ptr<column> concatenate_rows(
- * @throws cudf::logic_error if the input column is not at least two-level depth lists column (i.e.,
- *         each row must be a list of list).
+ * @throws std::invalid_argument if the input column is not at least two-level depth lists column
+ * (i.e., each row must be a list of list).
```
```suggestion
 *        (i.e., each row must be a list of list).
```

ttnghia commented on cpp/include/cudf/lists/contains.hpp:
```diff
@@ -163,7 +163,7 @@ std::unique_ptr<column> index_of(
  *
  * @throw cudf::logic_error If `search_keys` does not match `lists` in its number of rows
- * @throw cudf::logic_error If `search_keys` type does not match the element type in `lists`
+ * @throw cudf::dtype_error If `search_keys` type does not match the element type in `lists`
```
Please also remove the line below: it is no longer relevant.

ttnghia replied:
In addition, there is one more overload of `index_of` in this file that needs to be updated.

vyasr replied:
Is the line below no longer relevant because of changes in this PR, or because prior work converted this function to use the experimental comparators?

vyasr replied:
I updated the other overload as you suggested, thanks for catching that.

ttnghia commented on cpp/include/cudf/lists/gather.hpp:
```diff
@@ -43,7 +43,7 @@ namespace lists {
  *
  * @throws cudf::logic_error if `gather_map_list` size is not same as `source_column` size.
- * @throws cudf::logic_error if gather_map contains null values.
+ * @throws std::invalid_argument if gather_map contains null values.
```
Should the line below be updated with `dtype_error` (and the corresponding code)?

vyasr replied:
No, because nullability is not a property of a data type; a column of any data type could contain nulls. The new exception is specifically designed for cases when an operation is attempted on a data type that doesn't make sense, e.g. trying to compute the sqrt of a string column.

ttnghia commented on java/src/main/native/include/jni_utils.hpp:
```diff
@@ -861,6 +862,9 @@ inline void jni_cuda_check(JNIEnv *const env, cudaError_t cuda_status) {
     JNI_CHECK_CUDA_ERROR(env, cudf::jni::CUDA_ERROR_CLASS, e, ret_val);                            \
   }                                                                                                \
+  catch (const cudf::dtype_error &e) {                                                             \
+    JNI_CHECK_THROW_NEW(env, cudf::jni::CUDF_DTYPE_ERROR_CLASS, e.what(), ret_val);                \
```
Should be just `e` instead of `e.what()`? As the catch above did that.

vyasr commented:
@davidwendt I requested your review here since you had thoughts in the last meeting, let me know what you think of the changes in this PR. As you requested, I can rename `dtype_error` to `data_type_error` or so.

davidwendt commented on cpp/tests/strings/json_tests.cpp:
```diff
@@ -21,6 +21,7 @@
 
 #include <cudf_test/base_fixture.hpp>
 #include <cudf_test/column_wrapper.hpp>
+#include <stdexcept>
```
```suggestion

#include <stdexcept>
```

davidwendt commented on cpp/tests/lists/combine/concatenate_list_elements_tests.cpp:
```diff
@@ -20,6 +20,7 @@
 #include <cudf_test/type_lists.hpp>
 
 #include <cudf/lists/combine.hpp>
+#include <stdexcept>
```
```suggestion

#include <stdexcept>
```

davidwendt commented on cpp/tests/copying/segmented_gather_list_tests.cpp:
```diff
@@ -25,6 +25,7 @@
 #include <cudf/detail/null_mask.hpp>
 #include <cudf/lists/gather.hpp>
 #include <cudf/lists/lists_column_view.hpp>
+#include <stdexcept>
```
```suggestion

#include <stdexcept>
```

davidwendt commented on cpp/tests/copying/concatenate_tests.cu:
```diff
@@ -31,6 +31,7 @@
 
 #include <rmm/exec_policy.hpp>
 
+#include <stdexcept>
```
This include should move down with the other STL includes.

davidwendt commented on cpp/include/cudf/utilities/error.hpp:
```diff
@@ -86,6 +86,29 @@ struct fatal_cuda_error : public cuda_error {
 };
 /** @} */
```
The `/** @} */` should go after the struct `};` below.

davidwendt commented on cpp/include/cudf/utilities/error.hpp:
```diff
@@ -86,6 +86,29 @@ struct fatal_cuda_error : public cuda_error {
+  dtype_error(char const* const message) : std::invalid_argument(message) {}
+
+  /**
+   * @brief Construct a new logic error object with error message
```
```suggestion
   * @brief Construct a new data_type_error object with error message
```

davidwendt reviewed (changes requested):
I think `data_type_error` instead of `dtype_error` would fit better with libcudf naming.

davidwendt commented on cpp/include/cudf/utilities/error.hpp:
```diff
@@ -84,6 +84,29 @@ struct cuda_error : public std::runtime_error {
+ * @brief Exception thrown when an operation is attempted on an unsupported dtype.
+ *
+ * This exception should be thrown when an operation is attempted on an
+ * unsupported dtype. This exception should not be thrown directly and is
```
```suggestion
 * unsupported data_type. This exception should not be thrown directly and is
```

ttnghia commented on cpp/include/cudf/lists/contains.hpp:
```diff
@@ -129,6 +129,7 @@ enum class duplicate_find_option : int32_t {
  * @throw cudf::logic_error If `search_key` type does not match the element type in `lists`
  * @throw cudf::logic_error If `search_key` is of a nested type, or `lists` contains nested
  * elements (LIST, STRUCT)
+ * @throw cudf::data_type_error If `search_keys` type does not match the element type in `lists`
```
Please remove from line 129 to line 131.
Lines 130-131 are left from previous changes (now all overloads of `index_of` support nested types).

vyasr replied:
Done

ttnghia commented on cpp/include/cudf/lists/contains.hpp:
```diff
@@ -163,8 +164,8 @@ std::unique_ptr<column> index_of(
  * @throw cudf::logic_error If `search_keys` does not match `lists` in its number of rows
- * @throw cudf::logic_error If `search_keys` type does not match the element type in `lists`
  * @throw cudf::logic_error If `lists` or `search_keys` contains nested elements (LIST, STRUCT)
+ * @throw cudf::data_type_error If `search_keys` type does not match the element type in `lists`
```
Please remove line 167. It was left from previous changes, no longer relevant (`index_of` now supports nested types).

vyasr replied:
Done

vyasr commented:
/merge
